pkg/kubelet/server/stats/fs_resource_analyzer.go

```diff
@@ -0,0 +1,146 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package stats
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/types"
+	"k8s.io/kubernetes/pkg/volume"
+	"sync"
+	"time"
+)
+
+// PodVolumeStats encapsulates all VolumeStats for a pod
+type PodVolumeStats struct {
+	PodUid  types.UID
+	Volumes []VolumeStats
+}
+
+// fsResourceAnalyzerInterface is for embedding fs functions into ResourceAnalyzer
+type fsResourceAnalyzerInterface interface {
+	GetPodVolumeStats(uid types.UID) (PodVolumeStats, error)
+}
+
+// diskResourceAnalyzer provider stats about fs resource usage
+type fsResourceAnalyzer struct {
+	volumeCache         sync.RWMutex
+	statsProvider       StatsProvider
+	cachedVolumeStats   *cachedPodVolumeStats // Eagerly cache volume stats
+	calcVolumeFrequency time.Duration         // How frequently to calculate disk usage
+}
+
+var _ fsResourceAnalyzerInterface = &fsResourceAnalyzer{}
+
+// cachedPodVolumeStats caches the results of eagerly calculating PodVolumeStats
+type cachedPodVolumeStats struct {
+	volumeStats map[types.UID]*PodVolumeStats
+}
+
+// default calculation interval to use if none is provided
+const defaultVolumeStatsCalcInterval = 30 * time.Second
+
+// NewStats returns a new NewDiskResourceAnalyzer implementation
+func newFsResourceAnalyzer(statsProvider StatsProvider, calcVolumeFrequency time.Duration) *fsResourceAnalyzer {
+	return &fsResourceAnalyzer{
+		statsProvider:       statsProvider,
+		calcVolumeFrequency: calcVolumeFrequency,
+	}
+}
+
+// Start eager background caching of volume stats.
+func (s *fsResourceAnalyzer) Start() {
+	go func() {
+		sleepTime := s.calcVolumeFrequency
+		if sleepTime <= 0 {
+			glog.Warningf("volume stats calc interval must be greater than 0, defaulting to %v.", defaultVolumeStatsCalcInterval)
+			sleepTime = defaultVolumeStatsCalcInterval
+		}
+		for {
+			glog.V(3).Infof("Calculating volume stats and caching result.")
+			startTime := time.Now()
+			s.updateCachedPodVolumeStats()
+			elapsedDuration := time.Now().Sub(startTime)
+			glog.V(3).Infof("Finished calculating volume stats in %v, sleeping for %v.", elapsedDuration, sleepTime)
+			time.Sleep(sleepTime)
+		}
+	}()
+}
+
+// updateCachedPodVolumeStats calculates and caches the PodVolumeStats for every Pod known to the kubelet.
+func (s *fsResourceAnalyzer) updateCachedPodVolumeStats() {
+	// Calculate the new volume stats map
+	pods := s.statsProvider.GetPods()
+	newCache := &cachedPodVolumeStats{make(map[types.UID]*PodVolumeStats, len(pods))}
+	for _, pod := range pods {
+		podUid := pod.GetUID()
+		stats, err := s.getPodVolumeStats(podUid)
+		if err != nil {
+			continue
+		}
+		newCache.volumeStats[podUid] = &stats
```

Comment:
Storing non-pointer structs in maps has some disadvantages since the map value addresses are not constant.  I don't think we should store non-pointer structs in map values unless we have a strong reason to do so.  Can talk more offline